### PR TITLE
Staging

### PR DESCRIPTION
**Pull request by idanasulinmemphis**

### CODE REVIEW

File: examples/producer.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply examples/producer.py:1]: Fixed using Python prettier

File: examples/consumer.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        print("message: ", msg.get_data())
+        msg.ack()
+
+    def error_handler(error):
+        print("error: ", error)
+    try:
+        memphis = Memphis()
+        await memphis.connect(host="<memphis-host>", username="<application type username>", connection_token="<broker-token>")
+
+        consumer = await memphis.consumer(name="<station-name>", consumer_name="<consumer-name>", consumer_group="")
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply examples/consumer.py:17]: Fixed

File: memphis/memphis.py [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+            Exception: _description_
+        """
+        try:
+            await self.connection.broker_connection.publish(self.station_name + ".final", message, headers={
+                "Nats-Msg-Id": str(uuid.uuid4())
+            })
+        except Exception as e:
+            if hasattr(e, 'status_code') and e.status_code == '503':
+                raise Exception(
+                    "Produce operation has failed, please check wheether Station/Producer are still exist")
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:340]: Fixed

File: memphis/memphis.py [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+        """
+        try:
+            http_request("DELETE", 'http://'+self.connection.host+':'+str(self.connection.management_port)+'/api/consumers/destroyProducer', headers={
+                         "Authorization": "Bearer " + self.connection.access_token}, body_params={"name": self.producer_name, "station_name": self.station_name})
+        except:
+            return
+
+
+class Consumer:
+    def __init__(self, connection, station_name, consumer_name, consumer_group, pull_interval_ms, batch_size, batch_max_time_to_wait_ms, max_ack_time_ms):
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:355]: Done

File: memphis/memphis.py [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+        """Consume events.
+        """
+        try:
+            durable_name = self.consumer_group if self.consumer_group else self.consumer_name
+            self.psub = await self.connection.broker_connection.pull_subscribe(
+                self.station_name + ".final", durable=durable_name)
+            starttime = time.time()
+            while True:
+                try:
+                    time.sleep(self.pull_interval_ms/1000 -
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:376]: Fixed

File: memphis/memphis.py [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+                               ((time.time() - starttime) % self.pull_interval_ms/1000))
+                    msgs = await self.psub.fetch(self.batch_size)
+                    for msg in msgs:
+                        self.event.emit('message', Message(msg))
+                except Exception as e:
+                    self.event.emit('error', e)
+        except Exception as e:
+            raise Exception(e)
+
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:385]: https://docs.nats.io/using-nats/developer/develop_jetstream/streams

File: memphis/memphis.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+        self.data_port = 7766
+        self.username = None
+        self.connection_token = None
+        self.access_token_timeout = None
+        self.client = socket.socket()
+        self.reconnect_attempts = 0
+        self.reconnect = True
+        self.max_reconnect = 3
+        self.reconnect_interval_ms = 200
+        self.timeout_ms = 15000
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:28]: Fixed

File: memphis/memphis.py [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+
+        if data['ping_interval_ms']:
+            self.ping_interval_ms = data['ping_interval_ms']
+            t_ping_interval_ms = Thread(
+                target=self.__ping_server)
+            t_ping_interval_ms.daemon = True
+            t_ping_interval_ms.start()
+        if not self.connected:
+            try:
+                self.broker_manager = await broker.connect(servers=self.host+":"+str(self.data_port), allow_reconnect=True, reconnect_time_wait=2, connect_timeout=2, max_reconnect_attempts=60, token=self.connection_token)
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:97]: Fixed

File: memphis/memphis.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+                return Station(self, name.lower())
+            else:
+                raise Exception(e)
+
+    async def close(self):
+        """Close Memphis connection.
+        """
+        try:
+            if self.is_connection_active:
+                await self.broker_manager.close()
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:163]: Done

File: memphis/memphis.py [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+        self.reconnect = reconnect
+        self.max_reconnect = 9 if max_reconnect > 9 else max_reconnect
+        self.reconnect_interval_ms = reconnect_interval_ms
+        self.timeout_ms = timeout_ms
+        try:
+            self.client.connect((self.host, self.tcp_port))
+
+        except OSError as msg:
+            self.client = None
+            self.__close()
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:64]: Done

File: memphis/memphis.py [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        Args:
+            message (Uint8Array): message to send into the station.
+            ack_wait_sec (int, optional): max time in seconds to wait for an ack from memphis. Defaults to 15.
+
+        Raises:
+            Exception: _description_
+            Exception: _description_
+        """
+        try:
+            await self.connection.broker_connection.publish(self.station_name + ".final", message, headers={
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:334]: Fixed

File: memphis/memphis.py [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+                "Nats-Msg-Id": str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:335]: Fixed

File: memphis/memphis.py [hunk 13]
```diff
@@ -0,0 +1,408 @@
...
+        self.timeout_ms = timeout_ms
+        try:
+            self.client.connect((self.host, self.tcp_port))
+
+        except OSError as msg:
+            self.client = None
+            self.__close()
+
+        if self.client is None:
+            raise Exception("could not open socket")
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:67]: OK

File: .gitignore [hunk 14]
```diff
@@ -0,0 +1,160 @@
+# Byte-compiled / optimized / DLL files
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply .gitignore:4]: Removed

File: pyvenv.cfg [hunk 15]
```diff
@@ -0,0 +1,3 @@
+home = /Library/Frameworks/Python.framework/Versions/3.10/bin
+include-system-site-packages = false
+version = 3.10.2
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply pyvenv.cfg]: Removed

File: .gitignore (hunk 14 above)
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply .gitignore:4]: Added

File: memphis/memphis.py [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+        except Exception as e:
+            raise Exception(e)
+
+    async def consumer(self, station_name, consumer_name, consumer_group="", pull_interval_ms=1000, batch_size=10, batch_max_time_to_wait_ms=5000, max_ack_time_ms=30000, max_msg_deliveries=10):
+        """Creates a consumer.
+
+        Args:
+            station_name (str): station name to consume messages from.
+            consumer_name (str): name for the consumer.
+            consumer_group (str, optional): consumer group name. Defaults to "".
```
Comment (idanasulinmemphis) [inline memphis/memphis.py:258]: Defaults to the consumer name, fix on the docs as well

File: memphis/memphis.py [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+            message (Uint8Array): message to send into the station.
+            ack_wait_sec (int, optional): max time in seconds to wait for an ack from memphis. Defaults to 15.
+
+        Raises:
+            Exception: _description_
+            Exception: _description_
+        """
+        try:
+            await self.connection.broker_connection.publish(self.station_name + ".final", message, headers={
+                "Nats-Msg-Id": str(uuid.uuid4()), "ackWait": str(ack_wait_sec), "producedBy": self.producer_name})
```
Comment (idanasulinmemphis) [inline memphis/memphis.py:328]: I wonder if it works sending ackWait header

Comment (SvetaMemphis) [reply memphis/memphis.py:328]: Fixed

File: memphis/memphis.py [hunk 18]
```diff
@@ -0,0 +1,408 @@
...
+                target=self.__ping_server)
+            t_ping_interval_ms.daemon = True
+            t_ping_interval_ms.start()
+        if not self.connected:
+            try:
+                self.broker_manager = await broker.connect(servers=self.host+":"+str(self.data_port), allow_reconnect=True, reconnect_time_wait=2, connect_timeout=2, max_reconnect_attempts=60, token=self.connection_token)
+                self.broker_connection = self.broker_manager.jetstream()
+                self.connected = True
+            except Exception as e:
+                raise Exception(e)
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply memphis/memphis.py:94]: Done

File: examples/consumer.py [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+        consumer.event.on("message", msg_handler)
+        consumer.event.on("error", error_handler)
+        await consumer.consume()
```
[the remark this answers is not in the document]
Comment (SvetaMemphis) [reply examples/consumer.py:19]: couldn't find a better implementation

**Pull request closed without merging**